File: 44_Wildcard_Matching/Solution.py
class Solution:
    def match(self,s,p,i,j,n,m,dp):
        if j>=m:
            return i<n
        if i>=n:
            return (p[j]=='*' and self.match(s,p,i,j+1,n,m,dp))
        
        if dp.get(str(i)+','+str(j),None) != None:
            return dp.get(str(i)+','+str(j))
        else:
            if p[j] == '?' or p[j] == s[i]:
                dp[str(i)+','+str(j)] = True
                return self.match(s,p,i+1,j+1,n,m,dp)
            elif p[j] == '*':
                dp[str(i)+','+str(j)] = self.match(s,p,i,j+1,n,m,dp) or self.match(s,p,i+1,j+1,n,m,dp) or self.match(
                s,p,i+1,j,n,m,dp)
                return dp[str(i)+','+str(j)]
            else:
                dp[str(i)+','+str(j)] = False
                return False
        
    def match_(self, s, p, i, j, dp):
        if i == -1 and j == -1:
            return True
        if j == -1:
            return False
        if i == -1:
            return  p[j]=='*' and self.match_(s,p,i,j-1,dp)
        
        key = str(i)+","+str(j)
        
        if dp.get(key,None) != None:
            return dp.get(key)

        if s[i] == p[j] or p[j] == '?':
            dp[key] = self.match_(s,p,i-1,j-1,dp)            
        elif p[j] == '*':
            dp[key] = self.match_(s,p,i-1,j-1,dp) or self.match_(s,p,i-1,j,dp) or self.match_(s,p,i,j-1,dp)
        else:
            dp[key] = False
        return dp[key]

    def isMatch(self, s: str, p: str) -> bool:
        
        i = 0
        j = 0
        n = len(s)
        m = len(p)
        dp = {}
        return self.match(s,p,i,j,n,m,dp)


s = Solution()
s.isMatch("acdcb","a*c?b")

File: 44_Wildcard_Matching/Solution.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution{
    public:
        static bool f(int i, int j, int n, int m, vector<vector<char>> &dp, string &s, string &p){
            cout<<i<<" "<<j<<endl;
            if(i >=n){
                if(j>=m) return true;
                if(p[j] == '*') return f(i, j+1, n, m, dp, s, p);
                else return false;
            }
            if(j>=m) return false;
            if(dp[i][j] != -1) return dp[i][j]==0?false:true;

            bool t = false;
            if(s[i] == p[j]) t  = f(i+1, j+1, n, m, dp, s, p);
            else if(p[j] == '?') t = f(i+1, j+1, n, m, dp, s, p);
            else if(p[j] == '*') t = f(i+1, j, n, m, dp, s, p) || f(i+1, j+1, n, m, dp, s, p) || f(i, j+1, n, m, dp, s, p);
            else t = 0;
            dp[i][j] = t?1:0;
            return t;
        }

        static bool isMatch(string s, string p){
            int n = s.size(), m=p.size();
            vector<vector<char>> dp(n, vector<char>(m, -1));
            return f(0, 0, n, m, dp, s, p);
        }
};

int main(){
    string line;
    while(getline(cin, line)){
        int mid = line.find(' ');
        string s = line.substr(0, mid);
        string p = line.substr(mid+1, line.size()-mid-1);
        cout<<line<<endl;
        cout<<Solution::isMatch(s, p)<<endl<<endl;
    }
    return 0;
}
